jenkins/dataproc-custom-images/prepare-custom-leonardo-jupyter-dataproc-image.sh

```diff
@@ -0,0 +1,186 @@
+        "$(date +'%Y-%m-%dT%H:%M:%S%z')" \
+        "$*"
+}
+
+#
+# Main
+#
+log 'Installing prerequisites...'
+
+# Obtain the latest valid apt-key.gpg key file from https://packages.cloud.google.com to work
+# around intermittent apt authentication errors. See:
+# https://cloud.google.com/compute/docs/troubleshooting/known-issues
+retry 5 curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
+retry 5 apt-key update
+
+# install Docker
+# https://docs.docker.com/install/linux/docker-ce/debian/
+# export DOCKER_CE_VERSION="19.03.2~ce~3-0~debian"
+
+# retry 5 betterAptGet
+retry 5 apt-get install -y -q \
+    apt-transport-https \
+    ca-certificates \
+    curl \
+    gnupg2 \
+    software-properties-common
+
+log 'Adding Docker package sources...'
+
+apt-get remove docker docker-engine
+
+# shellcheck disable=SC1091
+os_dist_name="$(. /etc/os-release; echo "$ID")"
+os_dist_code_name="$(lsb_release -cs)"
+os_dist_release_channel="stable"
+os_dist_arch="amd64"
+
+docker_gpg_key_url="https://download.docker.com/linux/${os_dist_name:?}/gpg"
+docker_apt_repo_url="https://download.docker.com/linux/${os_dist_name:?}"
+
+retry 5 curl -fsSL "${docker_gpg_key_url:?}" | apt-key add -
+
+add-apt-repository \
+  "deb [arch=${os_dist_arch:?}] ${docker_apt_repo_url:?} \
+  ${os_dist_code_name:?} \
+  ${os_dist_release_channel:?}"
+
+log 'Installing Docker...'
+
+retry 5 apt-get update
+retry 5 add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"
+
+dpkg --configure -a
+#this line fails consistently, but it does not fail in a fatal way so we add `|| true` to prevent the script from halting execution
+apt-get install -y -q docker-ce || true
+
+log 'Installing Docker Compose...'
+
+# Install docker-compose
+# https://docs.docker.com/compose/install/#install-compose
+docker_compose_version_number="1.22.0"
+docker_compose_kernel_name="$(uname -s)"
+docker_compose_machine_hardware_name="$(uname -m)"
+docker_compose_binary_download_url="https://github.com/docker/compose/releases/download/${docker_compose_version_number:?}/docker-compose-${docker_compose_kernel_name:?}-${docker_compose_machine_hardware_name:?}"
+docker_compose_binary_download_target_filename="/usr/local/bin/docker-compose"
+
+retry 5 curl -L "${docker_compose_binary_download_url:?}" -o "${docker_compose_binary_download_target_filename:?}"
+chmod +x "${docker_compose_binary_download_target_filename:?}"
+
+# Pull docker image versions as of the time this script ran; this caches them in the
+# dataproc custom instance image.
+if [[ -n ${docker_image_var_names:?} ]]; then
+    for _docker_image_var_name in ${docker_image_var_names:?}
+    do
+        _docker_image="${!_docker_image_var_name:?}"
+        retry 5 docker pull "${_docker_image:?}"
+    done
+else
+    log "ERROR-VAR_NULL_OR_UNSET: docker_image_var_names. Will not pull docker images."
+fi
+
+# Install Python 3.6.8
+export PYTHON_VERSION="3.6.8"
```

Comment:
this needs to be updated to match the base image still I believe.

Reply:
Yeah we can probably `apt-get install` python 3.7 here, as we do in the base docker image.